Custom library-defined traits don't work with the codegen plugin

We might be doing something wrong, but it seems that custom library-defined traits are not properly supported by the codegen plugin.

**Our setup:**
- we have a library which defines some custom traits and registers the `TraitService` SPIs for them as well as some validators
- a service project is using the smithy4s codegen plugin & it depends on the above library as well as some JARs with smithy files

**The issue:**
When the model is loaded in the service project, it can't see the `TraitService` SPIs for our custom traits and it falls back to the dynamic model for those. Validators are picked-up fine, but they can't see the traits in the model since they rely on the runtime classes and won't work correctly with the dynamic model. Thus, we get validation errors everywhere. 

**The cause:**
The `ModelLoader` is only passing the classloader that can see external dependencies to the `discoverModels` call of the assembler and not to the factory method for the assembler itself:
```
Model
  .assembler() // this is using current classloader (which can only see the plugin's classpath)
  .discoverModels(upstreamClassLoader)
```
https://github.com/disneystreaming/smithy4s/blob/5ce1106e363bf1bec78e196791cbb26da6d803de/modules/codegen/src/smithy4s/codegen/ModelLoader.scala#L58
This way, it can't see any `TraitService` SPIs contributed by external dependencies, but it can still discover & load the smithy files contained in them.

**Our attempts to fix this:**
If we just pass the `upstreamClassloader` to the `Model.assembler` call, it loads the `TraitService` interface again from the `smithy-model` jar through this classloader and you end up in a situation where the SPI lookup fails with `"TraitService is not a TraitService"` type of error. We ran into similar kinds of issues when adding the current classloader as a parent of `upstreamClassLoader` and attempting to filter the external classpath.

It works if we abandon the staged model loading approach altogether and load everything from a unified classpath:
```
val validatorClassLoader = maybeDeps match {
  case Some(deps) => new URLClassLoader(deps, currentClassLoader)
  case None       => currentClassLoader
}

val modelAssembler =
  Model
    .assembler(validatorClassLoader)
    .discoverModels(validatorClassLoader) // discovering models from both internal & external classpath in a single pass
```
But that would surely break other things.

Update: still happening in 0.18.0

Having looked at the project @schmeedy was referring to _in the state at the time this was reported_, I think I found the particular corner case that reproduces this.

Consider this spec:

```smithy
namespace com.kubukoz

use alloy#simpleRestJson

@trait(selector: "service")
string apiVersion

@simpleRestJson
@apiVersion("v1")
service FooService {

}
```

If I define a trait service + validator... let's consider the validator doing something like this

1. Find all shapes with `@simpleRestJson`, for each of them do the following:
2. Ensure shape has `@apiVersion`
3. otherwise ERROR

then the validator doesn't see the real trait, but a dynamic default. At least that's what I'm seeing.

I'm aware that it's probably not a good pattern (adding behavior to a shape I don't own, `@simpleRestJson`) but that's long gone from the project and at this point I'm just reporting details in case we want to fix this.

Reproduction:

1. Clone https://github.com/kubukoz/demos/tree/smithy4s-validation-scope
2. `./mill demo.publishLocal`
3. `cs launch com.disneystreaming.smithy4s:smithy4s-codegen-cli_2.13:0.17.0 -- generate --dependencies com.kubukoz:demo:0.0.1-SNAPSHOT`

The output contains this:

```
software.amazon.smithy.model.validation.ValidatedResultException: Result contained ERROR severity validation events: 
[ERROR] com.kubukoz#FooService: Missing @apiVersion trait on service com.kubukoz#FooService. Found traits:

=======================
alloy#simpleRestJson: alloy.SimpleRestJsonTrait
com.kubukoz#apiVersion: software.amazon.smithy.model.traits.DynamicTrait
=======================
 | ApiVersioning jar:file:/Users/kubukoz/.ivy2/local/com.kubukoz/demo/0.0.1-SNAPSHOT/jars/demo.jar!/META-INF/smithy/demo.smithy:10:1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,10 @@ When adding entries, please treat them as if they could end up in a release any 
 
 Thank you!
 
+# 0.18.36
+
+* codegen: Pass the correct ClassLoader to prevent validators/transformers from breaking on externally-defined trait classes in [#1709](https://github.com/disneystreaming/smithy4s/pull/1709)
+
 # 0.18.35
 
 * json, documents: Add support for `@jsonUnknown` in unions (Open Unions) in [#1677](https://github.com/disneystreaming/smithy4s/pull/1677)

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/build.sbt
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/build.sbt
@@ -1,0 +1,23 @@
+lazy val transformation = project
+  .settings(
+    scalaVersion := "2.12.20",
+    libraryDependencies ++= Seq(
+      "software.amazon.smithy" % "smithy-build" % "1.57.1",
+      "ch.epfl.scala" % "spec-traits" % "2.2.0-M2"
+    )
+  )
+
+lazy val root = project
+  .in(file("."))
+  .enablePlugins(Smithy4sCodegenPlugin)
+  .settings(
+    libraryDependencies ++= Seq(
+      "ch.epfl.scala" % "spec-traits" % "2.2.0-M2" % Smithy4s,
+      "com.disneystreaming.smithy4s" %% "smithy4s-core" % smithy4sVersion.value
+    ),
+    Compile / smithy4sModelTransformers := List(
+      "my-transformation"
+    ),
+    Compile / smithy4sAllowedNamespaces := List("my.input"),
+    Compile / smithy4sAllDependenciesAsJars += (transformation / Compile / packageBin).value
+  )

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/build.sbt
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/build.sbt
@@ -11,6 +11,7 @@ lazy val root = project
   .in(file("."))
   .enablePlugins(Smithy4sCodegenPlugin)
   .settings(
+    scalaVersion := "3.3.6",
     libraryDependencies ++= Seq(
       "ch.epfl.scala" % "spec-traits" % "2.2.0-M2" % Smithy4s,
       "com.disneystreaming.smithy4s" %% "smithy4s-core" % smithy4sVersion.value

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/project/plugins.sbt
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/project/plugins.sbt
@@ -1,0 +1,9 @@
+sys.props.get("plugin.version") match {
+  case Some(x) =>
+    addSbtPlugin("com.disneystreaming.smithy4s" % "smithy4s-sbt-codegen" % x)
+  case _ =>
+    sys.error(
+      """|The system property 'plugin.version' is not defined.
+         |Specify this property using the scriptedLaunchOpts -D.""".stripMargin
+    )
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/src/main/scala/Main.scala
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/src/main/scala/Main.scala
@@ -1,0 +1,30 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+object Main extends App {
+  val hints = my.input.MyShape.schema.hints
+
+  require(
+    hints.has[smithy.api.Documentation],
+    s"Expected to have the Documentation trait, but it was missing: $hints"
+  )
+
+  require(
+    hints.get[smithy.api.Documentation].get.value == "what's up doc",
+    s"Documentation trait mismatch: ${hints.get[smithy.api.Documentation].get.value}"
+  )
+  println("all good: " + hints.all)
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/src/main/smithy/input.smithy
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/src/main/smithy/input.smithy
@@ -1,0 +1,6 @@
+$version: "2"
+
+namespace my.input
+
+@traits#data
+document MyShape

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/test
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/test
@@ -1,0 +1,4 @@
+# check if the app runs successfully.
+# if it doesn't compile, it could be because the transformation wasn't applied.
+# if the transformation throws, it's because of a bug like #336.
+> run

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/resources/META-INF/services/software.amazon.smithy.build.ProjectionTransformer
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/resources/META-INF/services/software.amazon.smithy.build.ProjectionTransformer
@@ -1,0 +1,1 @@
+MyTransformation

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/scala/MyTransformation.scala
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/scala/MyTransformation.scala
@@ -1,0 +1,43 @@
+import software.amazon.smithy.build.ProjectionTransformer
+import software.amazon.smithy.build.TransformContext
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.transform.ModelTransformer
+import java.util.function.BiFunction
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.traits.DocumentationTrait
+import software.amazon.smithy.model.shapes.ShapeId
+import bsp.traits.DataTrait
+
+class MyTransformation extends ProjectionTransformer {
+  def getName(): String = "my-transformation"
+
+  // Replace traits#jsonRPC with documentation
+  def transform(context: TransformContext): Model = {
+    // this would fail if the class wasn't present on the classpath.
+
+    // external shape - regression test for #336
+    context
+      .getModel()
+      .expectShape(ShapeId.from("bsp#BuildTargetData"))
+      .expectTrait(classOf[DataTrait])
+
+    // local shape
+    context
+      .getModel()
+      .expectShape(ShapeId.from("my.input#MyShape"))
+      .expectTrait(classOf[DataTrait])
+
+    ModelTransformer
+      .create()
+      .mapTraits(
+        context.getModel(),
+        {
+          case (_, _: DataTrait) =>
+            new DocumentationTrait("what's up doc")
+          case (_, trt) => trt
+        }: BiFunction[Shape, Trait, Trait]
+      )
+  }
+
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/scala/MyTransformation.scala
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/transformation-trait-classes/transformation/src/main/scala/MyTransformation.scala
@@ -1,3 +1,19 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
 import software.amazon.smithy.build.ProjectionTransformer
 import software.amazon.smithy.build.TransformContext
 import software.amazon.smithy.model.Model

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/build.sbt
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/build.sbt
@@ -1,0 +1,19 @@
+lazy val externalLibrary = project
+  .settings(
+    autoScalaLibrary := false,
+    crossPaths := false,
+    libraryDependencies ++= Seq(
+      "software.amazon.smithy" % "smithy-model" % "1.57.1"
+    )
+  )
+
+lazy val root = project
+  .in(file("."))
+  .enablePlugins(Smithy4sCodegenPlugin)
+  .settings(
+    scalaVersion := "3.3.6",
+    libraryDependencies ++= Seq(
+      "com.disneystreaming.smithy4s" %% "smithy4s-core" % smithy4sVersion.value
+    ),
+    Compile / smithy4sAllDependenciesAsJars += (externalLibrary / Compile / packageBin).value
+  )

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersionTrait.java
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersionTrait.java
@@ -1,3 +1,19 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
 import software.amazon.smithy.model.SourceLocation;
 import software.amazon.smithy.model.shapes.ShapeId;
 import software.amazon.smithy.model.traits.StringTrait;

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersionTrait.java
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersionTrait.java
@@ -1,0 +1,23 @@
+import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.StringTrait;
+
+public class ApiVersionTrait extends StringTrait {
+
+	public static ShapeId ID = ShapeId.from("my.input#apiVersion");
+
+	public ApiVersionTrait(String value, SourceLocation sourceLocation) {
+		super(ID, value, sourceLocation);
+	}
+
+	public ApiVersionTrait(String value) {
+		this(value, SourceLocation.NONE);
+	}
+
+	public static final class Provider extends StringTrait.Provider<ApiVersionTrait> {
+		public Provider() {
+			super(ID, ApiVersionTrait::new);
+		}
+	}
+
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersioningValidator.java
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersioningValidator.java
@@ -1,0 +1,22 @@
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public final class ApiVersioningValidator extends AbstractValidator {
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        model.getShapesWithTrait(ApiVersionTrait.ID).stream().forEach(shape -> {
+            System.out.println("validating shape: " + shape);
+            shape.expectTrait(ApiVersionTrait.class);
+        });
+
+        return List.of();
+    }
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersioningValidator.java
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/java/ApiVersioningValidator.java
@@ -1,3 +1,19 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
 import software.amazon.smithy.model.Model;
 import software.amazon.smithy.model.shapes.Shape;
 import software.amazon.smithy.model.validation.AbstractValidator;

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/services/software.amazon.smithy.model.traits.TraitService
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/services/software.amazon.smithy.model.traits.TraitService
@@ -1,0 +1,1 @@
+ApiVersionTrait$Provider

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/services/software.amazon.smithy.model.validation.Validator
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/services/software.amazon.smithy.model.validation.Validator
@@ -1,0 +1,1 @@
+ApiVersioningValidator

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/smithy/input.smithy
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/smithy/input.smithy
@@ -1,0 +1,9 @@
+$version: "2"
+
+namespace my.input
+
+@trait(selector: "service")
+string apiVersion
+
+@apiVersion("v1")
+service FooService {}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/smithy/manifest
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/externalLibrary/src/main/resources/META-INF/smithy/manifest
@@ -1,0 +1,1 @@
+input.smithy

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/project/plugins.sbt
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/project/plugins.sbt
@@ -1,0 +1,9 @@
+sys.props.get("plugin.version") match {
+  case Some(x) =>
+    addSbtPlugin("com.disneystreaming.smithy4s" % "smithy4s-sbt-codegen" % x)
+  case _ =>
+    sys.error(
+      """|The system property 'plugin.version' is not defined.
+         |Specify this property using the scriptedLaunchOpts -D.""".stripMargin
+    )
+}

--- a/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/test
+++ b/modules/codegen-plugin/src/sbt-test/codegen-plugin/validator-trait-classes/test
@@ -1,0 +1,3 @@
+# check if the app compiles.
+# if the validator throws, it's because of a bug like #336.
+> compile

--- a/modules/codegen-plugin/src/smithy4s/codegen/GenerateSmithyBuild.scala
+++ b/modules/codegen-plugin/src/smithy4s/codegen/GenerateSmithyBuild.scala
@@ -19,8 +19,9 @@ package smithy4s.codegen
 import sbt.Keys._
 import sbt._
 
-import Smithy4sCodegenPlugin.autoImport._
 import scala.collection.immutable.ListSet
+
+import Smithy4sCodegenPlugin.autoImport._
 
 private final case class SmithyBuildData(
     sources: ListSet[String],

--- a/modules/codegen-plugin/src/smithy4s/codegen/JsonConverters.scala
+++ b/modules/codegen-plugin/src/smithy4s/codegen/JsonConverters.scala
@@ -16,13 +16,14 @@
 
 package smithy4s.codegen
 
-import sjsonnew._
-import BasicJsonProtocol._
-import sbt.FileInfo
-import sbt.HashFileInfo
 import cats.data.Validated.Invalid
 import cats.data.Validated.Valid
+import sbt.FileInfo
+import sbt.HashFileInfo
 import sbt.io.Hash
+import sjsonnew._
+
+import BasicJsonProtocol._
 
 // Json codecs used by SBT's caching constructs
 private[smithy4s] object JsonConverters {

--- a/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
+++ b/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
@@ -76,7 +76,7 @@ private[codegen] object ModelLoader {
 
     // Loading the upstream model
     val upstreamModel = Model
-      .assembler()
+      .assembler(currentClassLoader)
       // disabling cache to support snapshot-driven experimentation
       .putProperty(ModelAssembler.DISABLE_JAR_CACHE, true)
       .addClasspathModels(currentClassLoader, discoverModels)

--- a/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
+++ b/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
@@ -74,9 +74,14 @@ private[codegen] object ModelLoader {
       }
     }
 
+    val validatorClassLoader = locally {
+      val jarUrls = deps.map(_.toURI().toURL()).toArray
+      new URLClassLoader(jarUrls, currentClassLoader)
+    }
+
     // Loading the upstream model
     val upstreamModel = Model
-      .assembler(currentClassLoader)
+      .assembler(validatorClassLoader)
       // disabling cache to support snapshot-driven experimentation
       .putProperty(ModelAssembler.DISABLE_JAR_CACHE, true)
       .addClasspathModels(currentClassLoader, discoverModels)
@@ -93,11 +98,6 @@ private[codegen] object ModelLoader {
       case (k, _) if k.startsWith("smithy4s") =>
         sanitisingModelBuilder.removeMetadataProperty(k)
       case _ => ()
-    }
-
-    val validatorClassLoader = locally {
-      val jarUrls = deps.map(_.toURI().toURL()).toArray
-      new URLClassLoader(jarUrls, currentClassLoader)
     }
 
     val preTransformationModel =

